x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Repeat.java

```diff
@@ -0,0 +1,107 @@
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+
+public class Repeat extends EsqlScalarFunction implements OptionalArgument {
+
+    private final Expression str;
+    private final Expression number;
+
+    @FunctionInfo(
+        returnType = "keyword",
+        description = "Returns a string constructed by concatenating `string` with itself the specified `number` of times.",
+        examples = @Example(file = "string", tag = "repeat")
+    )
+    public Repeat(
+        Source source,
+        @Param(name = "string", type = { "keyword", "text" }, description = "String expression.") Expression str,
+        @Param(name = "number", type = { "integer" }, description = "Number times to repeat.") Expression number
+    ) {
+        super(source, Arrays.asList(str, number));
+        this.str = str;
+        this.number = number;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.KEYWORD;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isString(str, sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return isType(number, dt -> dt == DataTypes.INTEGER, sourceText(), SECOND, "integer");
+    }
+
+    @Override
+    public boolean foldable() {
+        return str.foldable() && number.foldable();
+    }
+
+    @Evaluator
+    static BytesRef process(BytesRef str, int number) {
+        if (str == null) {
+            return null;
+        }
+        if (number < 0) {
+            throw new IllegalArgumentException("Number parameter cannot be negative, found [" + number + "]");
+        }
+
+        String utf8ToString = str.utf8ToString();
+        String repeated = utf8ToString.repeat(number);
+        return new BytesRef(repeated);
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new Repeat(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Repeat::new, str, number);
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        ExpressionEvaluator.Factory strExpr = toEvaluator.apply(str);
+        ExpressionEvaluator.Factory numberExpr = toEvaluator.apply(number);
```

Comment:
Optional, but probably most invocations would contain a foldable count (like in the tests). You could add an evaluator that no longer evaluates the count.

Reply:
That makes sense. I added a constant evaluator for that case. This section is the part of the PR that I feel most uncertain about it's correctness.